### ISSUE TITLE
Add overloads of Invokes, ReturnsLazily, etc for methods with more than 4 parameters

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1606]
As discussed [in chat](https://gitter.im/FakeItEasy?at=5cda88550f381d0a76b314b2) a while ago.

There are currently strongly-typed helper methods for `Invokes` for up to 4 parameters:

```csharp
.Invokes((T1 a) => ...)
.Invokes((T1 a, T2 b) => ...)
.Invokes((T1 a, T2 b, T3 c) => ...)
.Invokes((T1 a, T2 b, T3 c, T4 d) => ...)
```

The same is true for a few other configuration methods:
- `WhenArgumentsMatch`
- `ReturnsLazily`
- `AssignsOutAndRefParametersLazily`
- `Throws`
- `ThrowsAsync`

4 parameters is enough is most cases, but if you add an extra parameter, the build fails with a message that's not easily understood:

> CS1593 Delegate 'Action<IFakeObjectCall>' does not take 5 arguments

Even though it's bad practice to have methods too many parameters, it's not uncommon to have 5 or 6, so I think we should consider adding more overloads to support more parameter (maybe 8?).

These helper methods are a bit tedious to write, so we should probably consider using design-time code generation (T4 templates or another appropriate tool). If possible, the code generation could happen during build; if not, the code generation will be run when needed and the generated files will be committed.

**Comment by blairconrad** [comment]
For what it's worth, `Action` goes up to 16! 😈

**Comment by blairconrad** [comment]
We should consider tests. If we're going to generate them, then looking for build-time code generation make sense. If we're just going to test a few small parameter lists and the largest or whatever, I think manually-invoked code generation followed by commit is fine, as we'd likely be hand-editing tests any time we made a change.

**Comment by thomaslevesque** [comment]
> If we're going to generate them, then looking for build-time code generation make sense

Not necessarily. We can just run code generation manually for the code *and* the tests.

**Comment by blairconrad** [comment]
Sure. What I meant was that if we weren't going to generate the tests _at all_ (and I'm not saying we shouldn't, just if it proved problematic; not sure why it would), then spending extra effort looking for an at-build code generation scheme for the production code probably wouldn't pay—manually-run code generation would be sufficient.

Probably best to ignore all my comments. It'll simplify things.

**Comment by thomaslevesque** [comment]
Ah, I just remembered why I had left `Throws` and `AssignsOutAndRefParametersLazily`. The extension methods are no longer publicly exposed. They're just used as helpers for implementing the interface methods (We had to add the methods to the interfaces themselves, because otherwise the user would have had to explicitly specify the interface type in the type arguments, which would have been an awful experience...).

So, the code generation is going to be a bit uglier for those...
